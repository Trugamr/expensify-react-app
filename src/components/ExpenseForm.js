import React from 'react'

class ExpenseForm extends React.Component {
    state = {
        description: '',
        note: '',
        amount: 0
    }
    onDescriptionChange = (e) => {
        // e.persist() used to persist event, event is not available in set state function
        // due to react event pooling or simplay use a variable to store value first then update state
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onNoteChange = (e) => {
        e.persist()
        this.setState(() => ({ note: e.target.value }))
    }
    onAmountChange = (e) => {
        const regex = /^[0-9]*(\.[0-9]{0,2})?$/
        const amount = e.target.value
        if(regex.test(amount)) this.setState({ amount })        
    }
    render() {
        return (
            <div>
                <form>
                    <input
                        type="text"
                        placeholder="description"
                        autoFocus
                        value={ this.state.description }
                        onChange={ this.onDescriptionChange }
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={ this.state.amount }
                        onChange={ this.onAmountChange }
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={ this.state.note }
                        onChange={ this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm