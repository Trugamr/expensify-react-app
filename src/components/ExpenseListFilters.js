import React from 'react'
import { conect } from 'react-redux'
import { connect } from 'react-redux'
import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters'

const ExpenseListFilter = ({ filters, dispatch }) => (
    <div>
        <input type="text" value={filters.text} onChange={(e) => {
            dispatch(setTextFilter(e.target.value))
        }} />
        <select value={filters.sortBy} onChange={(e) => {
            if(e.target.value === 'date') {
                dispatch(sortByDate())
            } else if(e.target.value === 'amount') {
                dispatch(sortByAmount())
            }
        }}>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
        </select>
    </div>
)

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilter)