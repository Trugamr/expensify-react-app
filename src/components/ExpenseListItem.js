import React from 'react'
import { Link } from 'react-router-dom'
import moment from 'moment';
import numeral from 'numeral';

numeral.register('locale', 'in', {
    delimiters: {
        thousands: ',',
        decimal: '.'
    },
    currency: {
        symbol: 'â‚¹'
    }
})

numeral.locale('in')

const ExpressListItem = ({ id, description, amount, createdAt }) => (
    <div>
       <Link to={`/edit/${id}`}>
           <h3>{description}</h3>
        </Link>
       <p>
        {numeral(amount / 100).format('$0,0.00')} - {moment(createdAt).format('MMM Do, YYYY')}
       </p>
    </div>
)

export default ExpressListItem