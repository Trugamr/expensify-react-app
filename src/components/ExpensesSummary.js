import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from '../utils/numeral-locales';
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

numeral.locale('in')

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    const fomattedExpensestotal = numeral(expensesTotal/100).format('$0,0.00') 
    const expenseWord = expensesCount == 1 ? 'expense' : 'expenses'
    return (
        <div className="page-header">
            <div className="content-container">
                <div className="page_header__flex">
                    <h1 className="page-header__title">Viewing <span>{expensesCount}</span> {expenseWord} totalling <span>{fomattedExpensestotal}</span></h1>
                    <div className="page-header__action"><Link to="/create" className="button yellow">
                        <i className="fas fa-plus"></i> &nbsp;Add Expense
                    </Link></div>
                </div>
            </div>
        </div>
    )
}

const mapStateToProp = ({ expenses, filters }) => {
    const viewableEpenses = selectExpenses(expenses, filters)
    return {
        expensesCount: viewableEpenses.length,
        expensesTotal: selectExpensesTotal(viewableEpenses)
    }   
}

export default connect(mapStateToProp)(ExpensesSummary)