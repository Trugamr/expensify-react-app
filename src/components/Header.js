import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { startLogout } from '../actions/auth';

export const Header = (props) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard">
                    <img className="header__icon" src="/images/icon.png"></img><h1>Expensify</h1>
                </Link>
                <button className="button green" onClick={props.startLogout}>
                    Logout &nbsp;<i className="fas fa-sign-out-alt"></i>
                </button>
            </div>
        </div>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Header);