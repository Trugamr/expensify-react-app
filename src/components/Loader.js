import React from 'react'

export default class Loader extends React.Component {
    componentDidMount() {
        const loaderContainer = document.querySelector('.loader-container')
        lottie
            .loadAnimation({
                container: loaderContainer,
                renderer: 'svg',
                loop: true,
                autoplay: true,
                path: '/lottie-files/loader.json',
                rendererSettings: {
                    className: 'loader'
                }
            })
    }
    render() {
        return (
            <div className="loader-container">
            </div>
        )
    }
}