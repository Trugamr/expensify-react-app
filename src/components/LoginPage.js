import React from 'react'
import { connect } from 'react-redux'
import { startLogin } from '../actions/auth'

class LoginPage extends React.Component {
    constructor(props) {
        super(props)
    }
    render() {
        return (
            <div className="box-layout">
                <div className="box-layout__box">
                    <h1>Expensify</h1>
                    <p>managing expenses made easy</p>
                    <div className="box-layout__box__btn" onClick={this.props.startLogin}>
                        <i className="fab fa-google"></i>
                        <p>Login</p>
                    </div>
                </div>        
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage)