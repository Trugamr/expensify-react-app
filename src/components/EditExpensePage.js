import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from '../components/ExpenseForm'
import { startRemoveExpense, startEditExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        console.log(expense)
        this.props.startEditExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }
    onClick = () => {
            this.props.startRemoveExpense({ id: this.props.expense.id })
            this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                        // showing remove button on expense from
                        showRemove={true}
                    />
                    <button style={{display: 'none'}} className="remove-button button red" onClick={this.onClick}><i className="fas fa-trash"></i> &nbsp;Remove Expense</button>
                </div>
            </div>
        )
    }
}

// (state, props)
const mapStateToProps = ({ expenses }, { match }) => {    
    return {
        expense: expenses.find(expense => expense.id === match.params.id)
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)