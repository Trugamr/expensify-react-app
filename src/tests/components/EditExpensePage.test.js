import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'


let wrapper, history, startRemoveExpense, startEditExpense
beforeEach(() => {
    startRemoveExpense = jest.fn()
    startEditExpense = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(
        < EditExpensePage
            startEditExpense={startEditExpense}
            startRemoveExpense={startRemoveExpense}
            history={history}
            expense={expenses[0]}
        />
    )
})

test('should render edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle startEditExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0])    
})

test('should hande startRemoveExpense', () => {
    wrapper.find('button').simulate('click')
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[0].id })
})