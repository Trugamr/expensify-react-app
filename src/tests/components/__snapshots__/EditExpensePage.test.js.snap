// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <Connect(ExpenseForm)
      expense={
        Object {
          "amount": 120,
          "createdAt": 0,
          "description": "Bubble Gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
      showRemove={true}
    />
    <button
      className="remove-button button red"
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
    >
      <i
        className="fas fa-trash"
      />
       Â Remove Expense
    </button>
  </div>
</div>
`;
